include/xtensor/xoperation.hpp

```diff
@@ -81,6 +81,43 @@ namespace xt
         }                                                              \
     }
 
+// This functor avoids implicit conversions of small integral types to 'int'
+// by returning the same type instead of 'auto'.
+#define UNARY_BITWISE_OPERATOR_FUNCTOR(NAME, OP)                   \
+    struct NAME                                                    \
+    {                                                              \
+        template <class A1>                                        \
+        constexpr std::decay_t<A1> operator()(const A1& arg) const \
+        {                                                          \
+            return OP arg;                                         \
+        }                                                          \
+        template <class B>                                         \
+        constexpr auto simd_apply(const B& arg) const              \
+        {                                                          \
+            return OP arg;                                         \
+        }                                                          \
+    }
+
+// This functor avoids implicit conversions of small integral types to 'int'
+// by returning the largest type instead of 'auto'.
+#define BINARY_BITWISE_OPERATOR_FUNCTOR(NAME, OP)                                                                    \
+    struct NAME                                                                                                      \
+    {                                                                                                                \
+        template <class T1, class T2>                                                                                \
+        constexpr std::                                                                                              \
+            conditional_t<(sizeof(std::decay_t<T1>) > sizeof(std::decay_t<T2>)), std::decay_t<T1>, std::decay_t<T2>> \
+            operator()(T1&& arg1, T2&& arg2) const                                                                   \
+        {                                                                                                            \
+            using xt::detail::operator OP;                                                                           \
+            return (std::forward<T1>(arg1) OP std::forward<T2>(arg2));                                               \
+        }                                                                                                            \
+        template <class B>                                                                                           \
```

Comment:
Why don't you allow the two types here?

Reply:
The existing `BINARY_OPERATOR_FUNCTOR` also has a single template argument for `simd_apply`. I don't know why. Using two arguments seems more logical, indeed.